extensions/amp-bind/amp-bind.md

```diff
@@ -66,425 +72,972 @@ The `amp-bind` component allows you to add custom stateful interactivity to your
+
+#### Removing a variable
+
+Remove an existing state variable by setting its value to `null` in `AMP.setState()`.
+
+```html
+<button on="tap:AMP.setState({removeMe: null})"></button>
+```
+
+#### Deep-merge with `AMP.setState()`
+
+Calling `AMP.setState()` deep-merges the provided object literal with the current state. `amp-bind` writes all literals to the state directly, except for nested objects, which are recursively merged. Primitives and arrays are in the state are always overwritten by variables of the same name in the object literal.
+
+[example preview="inline" playground="true" imports="amp-bind"]
+
+```html
+<p [text]="employee.name">Name</p>
+<p [text]="employee.age">Age</p>
+<p [text]="employee.vehicle">Vehicle</p>
+<!-- Pressing this button changes state to: -->
+<button
+  on="tap:AMP.setState({
+              employee: {
+                name: 'John Smith',
+                age: 47,
+                vehicle: 'Car'
+              }
+            })"
+>
+  Set employee to John Smith
+</button>
+<!-- Pressing this button recursively merges the object literal argument,  -->
+<!-- `{employee: {age: 64}}`, into the existing state. -->
+
+<button
+  on="tap:AMP.setState({
+              employee: {
+                age: 64
+              }
+            })"
+>
+  Set employee age to 64
+</button>
+<!-- The value updates from 47 to 64 at employee.age.  -->
+<!-- No other values change. -->
+```
+
+[/example]
+
+### Modifying history with `AMP.pushState()`
+
+`AMP.pushState()` writes state changes to the history. Navigating back, will restore the previous state. To test this, increase the count in the example below and use your browser's back button to decrease the count.
+
+[example preview="inline" playground="true" imports="amp-bind"]
+
+```html
+<amp-state id="count">
+  <script type="application/json">
+    1
+  </script>
+</amp-state>
+<div>Item <span [text]="count">1</span></div>
+<button on="tap:AMP.pushState({ count: count + 1 })">Increase count</button>
+```
+
+[/example]
+
+Using `AMP.pushState()` sets the current state to the most recent pushed state.
+
+## Expressions
+
+`amp-bind` uses JavaScript-like expressions that can reference the state.
+
+### Differences from JavaScript
+
+- Expressions may only access the containing document's [state](#state).
+- Expressions **do not** have access to `window` or `document`. `global` references the top-level state.
+- Only `amp-bind` [allowed-listed functions](#allowed-listed functions) and operators are usable. are usable. Use of arrow functions are allowed as function parameters, e.g. `[1, 2, 3].map(x => x + 1)`.
+  - Custom functions, classes and loops are disallowed.
+- Undefined variables and array-index-out-of-bounds return `null` instead of `undefined` or throwing errors.
+- A single expression is currently capped at 50 operands for performance. Please [contact us](https://github.com/ampproject/amphtml/issues/new) if this is insufficient for your use case.
```

Comment:
I thought this limit was recently upped. Might be worth coming back validating in a followup PR.

Reply:
Noted! Thank you.